pkg/controller/deadmanssnitchintegration/deadmanssnitchintegration_controller.go

```diff
@@ -0,0 +1,530 @@
+	}
+
+	// Watch for changes to Secrets. If one has any ClusterDeployment owner
+	// references, queue a request for all DeadMansSnitchIntegration CR that
+	// select those ClusterDeployments.
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: ownedByClusterDeploymentToDeadMansSnitchIntegrationsMapper{
+				Client: mgr.GetClient(),
+			},
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileDeadmansSnitchIntegration implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileDeadmansSnitchIntegration{}
+
+// ReconcileDeadmansSnitchIntegration reconciles a DeadmansSnitchIntegration object
+type ReconcileDeadmansSnitchIntegration struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client    client.Client
+	scheme    *runtime.Scheme
+	dmsclient func(apiKey string, collector *localmetrics.MetricsCollector) dmsclient.Client
+}
+
+// Reconcile reads that state of the cluster for a DeadmansSnitchIntegration object and makes changes based on the state read
+// and what is in the DeadmansSnitchIntegration.Spec
+func (r *ReconcileDeadmansSnitchIntegration) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling DeadmansSnitchIntegration")
+	// Fetch the DeadmansSnitchIntegration dmsi
+	dmsi := &deadmanssnitchv1alpha1.DeadmansSnitchIntegration{}
+
+	start := time.Now()
+	defer func() {
+		reconcileDuration := time.Since(start).Seconds()
+		reqLogger.WithValues("Duration", reconcileDuration).Info("Reconcile complete.")
+		localmetrics.Collector.ObserveReconcile(reconcileDuration)
+	}()
+
+	err := r.client.Get(context.TODO(), request.NamespacedName, dmsi)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	dmsAPIKey, err := utils.LoadSecretData(r.client, dmsi.Spec.DmsAPIKeySecretRef.Name,
+		dmsi.Spec.DmsAPIKeySecretRef.Namespace, DeadMansSnitchAPISecretKey)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	dmsc := r.dmsclient(dmsAPIKey, localmetrics.Collector)
+
+	matchingClusterDeployments, err := r.getMatchingClusterDeployment(dmsi)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if dmsi.DeletionTimestamp != nil {
+		for _, clustDeploy := range matchingClusterDeployments.Items {
+			err = r.deleteDMSI(dmsi, &clustDeploy, dmsc)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+		}
+		return reconcile.Result{}, nil
```

Comment:
```suggestion
		logger.Info("Deleting DMS finalizer from dmsi")
		if utils.HasFinalizer(dmsi, DeadMansSnitchFinalizer) {
			utils.DeleteFinalizer(dmsi, DeadMansSnitchFinalizer)
			err = r.client.Update(context.TODO(), dmsi)
			if err != nil {
				logger.Error(err, "Error deleting Finalizer from dmsi")
				return reconcile.Result{}, err
			}
		}

		return reconcile.Result{}, nil
```

Only delete DSMI finalizer after done deleting DMS services for each CD.

Reply:
You only want things to be deleted when the Clusterdeployment is deleted ? Not when the dms integration cr is deleted?

Reply:
@R-Lawton I _think_ what @jewzaam is suggesting here should be taken in the context of the previous suggestion on L156, but there might be some info missing between the two:

`r.deleteDMSI` and `r.deleteDMSClusterDeploy` are essentially duplicates of each other, except one deletes the finalizer on the DMSI and the other deletes the finalizer on the ClusterDeployment. When we're cleaning stuff up here and deleting the finalizer from the DMSI, the relevant finalizer should also be removed from the ClusterDeployment. So changing line 156 to call `r.deleteDMSClusterDeploy`, and then adding the removal of the finalizer from the DMSI here (outside the range over corresponding ClusterDeployments, since it only needs to be removed once), means that the `r.deleteDMSI` method can be removed.

Does that sound reasonable/correct? @jewzaam please correct me if you were suggesting something else. :slightly_smiling_face:

Reply:
Yea, without getting into what function does what just need to be sure this does what is expected.
1. if CD is deleted, DMS stuff is cleaned up before finalizer is removed from CD
2. if DMSI is deleted, each CD is cleaned up with finalzier removed from CD before finalizer is removed from DMSI

Reply:
Perfect sounds good im on the same page